File: word-wrap/csharp/src/WordWrap/WordWrap.cs
using System.Text;

namespace WordWrap;

public static class WordWrap
{
    public static string Wrap(string text, int width)
    {
        var words = text.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
        if (words.Length == 0)
        {
            return string.Empty;
        }

        var lines = new List<string>();
        var line = new StringBuilder();

        foreach (var original in words)
        {
            var word = original;
            while (word.Length > width)
            {
                if (line.Length > 0)
                {
                    lines.Add(line.ToString());
                    line.Clear();
                }
                lines.Add(word[..width]);
                word = word[width..];
            }

            if (line.Length == 0)
            {
                line.Append(word);
            }
            else if (line.Length + 1 + word.Length <= width)
            {
                line.Append(' ').Append(word);
            }
            else
            {
                lines.Add(line.ToString());
                line.Clear();
                line.Append(word);
            }
        }

        if (line.Length > 0)
        {
            lines.Add(line.ToString());
        }

        return string.Join('\n', lines);
    }
}

File: word-wrap/python/src/word_wrap/word_wrap.py
from __future__ import annotations


def wrap(text: str, width: int) -> str:
    words = text.split()
    if not words:
        return ""

    lines: list[str] = []
    line = ""

    for original in words:
        word = original
        while len(word) > width:
            if line:
                lines.append(line)
                line = ""
            lines.append(word[:width])
            word = word[width:]

        if not line:
            line = word
        elif len(line) + 1 + len(word) <= width:
            line = line + " " + word
        else:
            lines.append(line)
            line = word

    if line:
        lines.append(line)

    return "\n".join(lines)
